plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -311,89 +396,192 @@ func (c *CloudWatch) fetchNamespaceMetrics() ([]*cloudwatch.Metric, error) {
-		}
-		if point.Minimum != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticMinimum)] = *point.Minimum
-		}
-		if point.SampleCount != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticSampleCount)] = *point.SampleCount
-		}
-		if point.Sum != nil {
-			fields[formatField(*metric.MetricName, cloudwatch.StatisticSum)] = *point.Sum
+				for j := range result.Values {
+					if _, ok := thing[*result.Timestamps[j]]; !ok {
+						thing[*result.Timestamps[j]] = metric{fields: map[string]interface{}{*result.Label: *result.Values[j]}, tags: tags}
+					}
+					thing[*result.Timestamps[j]].fields[*result.Label] = *result.Values[j]
+				}
+			}
 		}
 
-		acc.AddFields(formatMeasurement(c.Namespace), fields, tags, *point.Timestamp)
+		for t, m := range thing {
+			acc.AddFields(namespace, m.fields, m.tags, t)
+		}
 	}
 
 	return nil
 }
 
+func nameMatch(name, id string) bool {
+	if strings.TrimPrefix(id, "average_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "maximum_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "minimum_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "sum_") == name {
+		return true
+	}
+	if strings.TrimPrefix(id, "sample_count_") == name {
+		return true
+	}
+	return false
+}
+
 /*
  * Formatting helpers
  */
-func formatField(metricName string, statistic string) string {
-	return fmt.Sprintf("%s_%s", snakeCase(metricName), snakeCase(statistic))
-}
-
 func formatMeasurement(namespace string) string {
 	namespace = strings.Replace(namespace, "/", "_", -1)
 	namespace = snakeCase(namespace)
-	return fmt.Sprintf("cloudwatch_%s", namespace)
+	return "cloudwatch_" + namespace
 }
 
 func snakeCase(s string) string {
 	s = internal.SnakeCase(s)
+	s = strings.Replace(s, " ", "_", -1)
 	s = strings.Replace(s, "__", "_", -1)
 	return s
 }
 
-/*
- * Map Metric to *cloudwatch.GetMetricStatisticsInput for given timeframe
- */
-func (c *CloudWatch) getStatisticsInput(metric *cloudwatch.Metric) *cloudwatch.GetMetricStatisticsInput {
-	input := &cloudwatch.GetMetricStatisticsInput{
-		StartTime:  aws.Time(c.windowStart),
-		EndTime:    aws.Time(c.windowEnd),
-		MetricName: metric.MetricName,
-		Namespace:  metric.Namespace,
-		Period:     aws.Int64(int64(c.Period.Duration.Seconds())),
-		Dimensions: metric.Dimensions,
-		Statistics: []*string{
-			aws.String(cloudwatch.StatisticAverage),
-			aws.String(cloudwatch.StatisticMaximum),
-			aws.String(cloudwatch.StatisticMinimum),
-			aws.String(cloudwatch.StatisticSum),
-			aws.String(cloudwatch.StatisticSampleCount)},
-	}
-	return input
+type queryData struct {
+	metric *cloudwatch.Metric
+	id     string
+}
+
+// get all of the possible queries so we can send groups of 100
+func (c *CloudWatch) getDataQueries(filteredMetrics []filteredMetric) ([]*cloudwatch.MetricDataQuery, error) {
+	if c.queryCache != nil && c.metricCache != nil && c.metricCache.IsValid() {
+		return c.queryCache, nil
+	}
+
+	// todo: lock
```

Comment:
Does this todo need to happen?

Reply:
I guess not, as it's used asynchronously